src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,204 @@
+#include "scheduling_alg.h"
+#include <iostream>
+#include <stdlib.h>
+#include <chrono>
+
+using namespace std;
+
+ClusterResources::ClusterResources(int64_t local_node_id, const NodeResources &node_resources) {
+  local_node_id_ = local_node_id;
+  add(local_node_id, node_resources);
+}
+
+void ClusterResources::add(int64_t node_id, const NodeResources &node_resources) {
+
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    // This node is new. Add it to the list.
+    nodes_.emplace(node_id, node_resources);
+  } else {
+    // This node exists. Update it's resources.
+    NodeResources &nr = it->second;
+
+    // First, update its predefined resources.
```

Comment:
Would be nice to make this a utility function: `NodeResources::SetPredefinedResources`.